sys/freebsd/pf.txt

```diff
@@ -80,6 +81,18 @@ ioctl$DIOCIGETIFACES(fd fd_pf, cmd const[DIOCIGETIFACES], arg ptr[in, pfioc_ifac
 ioctl$DIOCSETIFFLAG(fd fd_pf, cmd const[DIOCSETIFFLAG], arg ptr[in, pfioc_iface])
 ioctl$DIOCCLRIFFLAG(fd fd_pf, cmd const[DIOCCLRIFFLAG], arg ptr[in, pfioc_iface])
 ioctl$DIOCKILLSRCNODES(fd fd_pf, cmd const[DIOCKILLSRCNODES], arg ptr[in, array[int8]])
+ioctl$DIOCGIFSPEEDV0(fd fd_pf, cmd const[DIOCGIFSPEEDV0], arg ptr[in, pf_ifspeed_v0])
+ioctl$DIOCGIFSPEEDV1(fd fd_pf, cmd const[DIOCGIFSPEEDV1], arg ptr[in, pf_ifspeed_v1])
+ioctl$DIOCGETSTATESV2(fd fd_pf, cmd const[DIOCGETSTATESV2], arg ptr[in, pfioc_states_v2])
+ioctl$DIOCGETSYNCOOKIES(fd fd_pf, cmd const[DIOCGETSYNCOOKIES], arg ptr[in, array[int8]])
```

Comment:
Shouldn't some of these pointers have `out` or `inout` annotations instead, for the case where the kernel returns some data to the caller?

Reply:
I don't know what these ioctl's are doing, but from the fuzzer perspective:
 - `out` is important for structs that contain resource, so that the fuzzer will take them out of the returned structs and use in other syscalls
 - `out` also prevents the fuzzer from fill-in the data and spending cycles mutating it


Now we also have notation to mark direction per-field, e.g.:
```
foo {
    a    int32 (in)
    b    int32 (out)
    c    int32 # will use the pointer direction by default
}
```

Reply:
> This looks a bit suspicious:
> 
> ```
> ioctl$DIOCGETSYNCOOKIES(fd fd_pf, cmd const[DIOCGETSYNCOOKIES], arg ptr[in, array[int8]])
> ```
> 
> we pass a variable-size array, but don't pass in the size, so how can kernel possibly know the size? Maybe it should be something like:
> 
You are correct. I messed up those calls. They're part of a group of new ioctl that take nvlist arguments, but those all pass a `struct pfioc_nv { void *date; size_t len; size_t size}`, which communicates the size of the nvlist as well as the available memory for a reply from the kernel.

I'll fix that.